fix: validate private double spends in txs with public funcs

natebeauregard opened the pull request:
Fixes https://github.com/AztecProtocol/aztec-packages/issues/12902

The post-processing double-spend validation broke once the PublicTxSimulator+AVM started performing all nullifier insertions.

This PR removes the post-processing double spend validation and instead checks all txs (regardless of if the tx is private-only or has public calls) for duplicate nullifiers during pre-processing. Then, the PublicTxSimulator+AVM will ensure that any nullifiers added during public execution cannot cause collisions.

see https://demerzelsolutions.slack.com/archives/C04BTJAA694/p1743001880033519 for additional info

natebeauregard committed:
fix: validate private double spends in txs with public funcs

dbanks12 reviewed (commented):
Ah interesting that the only post-processing validation we were doing was checking double spends. @spalladino do you think it's alright to completely remove post-processing validation like this?

natebeauregard committed:
chore: update double spend unit test case

spalladino reviewed (approved):
Looks good, but let's add a test that catches this somehow. I'm fine if we have to go all the way up to an e2e test, but we should assert that:
- If we send a tx that emits a duplicate nullifier in public app logic, it is mined and reverts
- If we send a tx that emits a duplicate nullifier in public setup logic, it is not included

It's possible there are already e2e tests for that, but let's double check. Doesn't need to be on this PR btw.

natebeauregard commented:
> Looks good, but let's add a test that catches this somehow. I'm fine if we have to go all the way up to an e2e test, but we should assert that:
> 
> * If we send a tx that emits a duplicate nullifier in public app logic, it is mined and reverts
> * If we send a tx that emits a duplicate nullifier in public setup logic, it is not included
> 
> It's possible there are already e2e tests for that, but let's double check. Doesn't need to be on this PR btw.

@spalladino Good idea, I think that this is probably important enough to include in this PR before merging to ensure that the full flow is working as expected.

> If we send a tx that emits a duplicate nullifier in public app logic, it is mined and reverts

I noticed that there was an e2e test ([e2e_private_voting_contract](https://github.com/AztecProtocol/aztec-packages/blob/1e58eb1511e01463ba8f7052839b1f0d85e6f2ef/yarn-project/end-to-end/src/e2e_private_voting_contract.test.ts)) that should already handle this case. However, my changes above cause the assertion to see the tx reject throw with `APP_LOGIC_REVERTED` tx status fail and instead just throws the basic `"Transaction 0x0b5eac8b283d70c750f0b777df3f0b921daba41694926145b861ad8ba597073b was dropped. Reason: Tx dropped by P2P node"` message. I'm currently debugging this to figure out what's going on.

> If we send a tx that emits a duplicate nullifier in public setup logic, it is not included

I can add a new e2e test case to assert this behavior.

natebeauregard committed:
chore: update existing double spend tests

natebeauregard committed:
chore: add double spend e2e test

natebeauregard commented:
I added commits to update the existing unit/e2e tests and add in a new e2e test for txs that emit a duplicate nullifier in public app logic. After talking with @spalladino , he suggested that we skip adding the e2e test in public setup logic for now and split that out into a separate issue.

I didn't add the new e2e test to CI yet since external contributors can't merge in CI changes, however I verified that the test passed locally. @spalladino @dbanks12 Would either of you be willing to make a small PR once this goes in to add the new test to [CI](https://github.com/AztecProtocol/aztec-packages/blob/master/yarn-project/end-to-end/bootstrap.sh)? Thanks in advance.
